plugins_src/import_export/wpc_jscad.erl

```diff
@@ -0,0 +1,189 @@
+%%
+%%     $Id$
+%%
+
+-module(wpc_jscad).
+
+-export([init/0,menu/2,command/2]).
+
+-include_lib("wings/e3d/e3d.hrl").
+-include_lib("wings/intl_tools/wings_intl.hrl").
+
+init() ->
+    true.
+
+menu({file,export}, Menu) ->
+    menu_entry(Menu);
+menu({file,export_selected}, Menu) ->
+    menu_entry(Menu);
+menu(_, Menu) -> Menu.
+
+command({file,{export,{jscad,Ask}}}, St) ->
+    Exporter = fun(Ps, Fun) -> wpa:export(Ps, Fun, St) end,
+    do_export(Ask, export, Exporter, St);
+command({file,{export_selected,{jscad,Ask}}}, St) ->
+    Exporter = fun(Ps, Fun) -> wpa:export_selected(Ps, Fun, St) end,
+    do_export(Ask, export_selected, Exporter, St);
+command(_, _) ->
+    next.
+
+menu_entry(Menu) ->
+    Menu ++ [{"OpenJSCAD (.jscad)...",jscad,[option]}].
+
+props() ->
+    [{ext,".jscad"},{ext_desc,?__(1,"OpenJSCAD File")}].
+
+%%%
+%%% Export.
+%%%
+
+do_export(Ask, Op, _Exporter, _St) when is_atom(Ask) ->
+    wpa:dialog(Ask, ?__(1,"OpenJSCAD Export Options"), dialog(export),
+	       fun(Res) ->
+		   {file,{Op,{jscad,Res}}}
+	       end);
+do_export(Attr, _Op, Exporter, _St) when is_list(Attr) ->
+    set_pref(Attr),
+    SubDivs = proplists:get_value(subdivisions, Attr, 0),
+    Ps = [{subdivisions,SubDivs}|props()],
+    Exporter(Ps, export_fun(Attr)).
+
+export_fun(Attr) ->
+    fun(Filename, Contents) ->
+	export(Filename, Contents, Attr)
+    end.
+
+export(Filename, Contents0, Attr) ->
+    Contents = export_transform(Contents0, Attr),
+    #e3d_file{objs=Objs,mat=Mat,creator=Creator} = Contents,
+    {ok,F} = file:open(Filename, [write]),
+    FName = filename:basename(Filename)++filename:extension(Filename),
+    NObjs = length(Objs),
+    %% Write file head
+    io:format(F, "// File         : ~ts\n",[FName]),
+    io:format(F, "// Objects      : ~lp\n",[NObjs]),
+    io:format(F, "// Exported from: ~ts\n",[Creator]),
+    %% Write file body
+    io:format(F, "function main() {\n",[]),
+    io:format(F, "  return [\n",[]),
+    try
+	lists:foldl(fun(#e3d_object{name = Name, obj=Obj}, AccObj) ->
+			Meshes = e3d_mesh:split_by_material(Obj),
+			NMeshes = length(Meshes),
+			all(fun(Mesh,AccMeshes) ->
+				export_object(F, Name, Mesh, Mat),
+				AccMeshes-1
+			    end, F, NMeshes, Meshes),
+			if AccObj > 1 ->
+			    io:put_chars(F, ",\n");
+			    true -> ignore
+			end,
+			AccObj-1
+		  end, NObjs, Objs),
+    	io:format(F, "\n  ];\n}\n",[])
+    catch _:Err:Stacktrace ->
+	io:format(?__(1,"OpenJSCAD Error: ~P in")++" ~p~n", [Err,30,Stacktrace])
+    end,
+    ok = file:close(F).
+
+export_object(F, ObjName, #e3d_mesh{type=Type,fs=Fs,vs=VTab}, Mat_defs) ->
+    [#e3d_face{mat=[Material|_]}|_] = Fs,
+    DifColor = material(Material, Mat_defs),
+    io:format(F, "    // ~ts.~ts\n",[ObjName,Material]),
+    io:format(F, "    color(~p, \n",[DifColor]),
+    io:format(F, "      polyhedron(\n",[]),
+
+    %% Helpers
+    W3 = fun({X,Y,Z}) ->  io:format(F, "                    [~p,~p,~p]", [X,Y,Z]) end,
```

Comment:
Use  `"\t\t[~f,~f,~f]"`  instead ?

Reply:
I avoid to use "\t" (and used two spaces instead) because I wanted to get the file readable. With six indentation level the line would become long for evaluation in the editor

Reply:
I didn't want "\t\t\t\t\t"  I wanted "\t" instead of 8 spaces :-)

Reply:
It looked much better to read the source in the OpenJSCAD editor. But, no problem, you are the boss. :D